If the build step in tekton pipeline fails, tekton dashboard shows the push step as green

# Expected Behavior

- When the `build-bud` step in kabanero pipeline fails due to some reason, the `build-push` step never runs from kabanero pipelines perspective.
And same should be seen on the Tekton-Dashboard with red mark for `build-bud` step and `build-push` step.

# Actual Behavior

- When the `build-bud` step is failing the the pipelinerun, still the Tekton-Dashboard is showing that `build-push` step succeeded.

# Steps to Reproduce the Problem

1.  Create a webhook for java-microprofile project from the tekton dashboard.
2. Make some code changes with compile time errors , so the `build-bud` step fails.


# Additional Info

   From Kabanero pipelines perspective we created webhook for java-microprofile project in my github.ibm.com from the tekton dashboard.
    I made some code changes with compile time errors , so the `build-bud` step fails.

Results:

    

- As the pipelinerun failed, the tekton dashboard shows below screenshot.

<img width="1374" alt="Screen Shot 2019-10-15 at 1 05 31 PM" src="https://user-images.githubusercontent.com/53223993/66858310-036a4480-ef57-11e9-9831-e1d79c72eeab.png">



-  However when we investigated from the oc logs command we found below results



- Logs Snippet for build-bud container in the pod (which was as expected to build fail):-

`oc logs -f my-java-mp-webhook-1571152644-build-task-wndcp-pod-2f0e7d -c step-build-bud`
```
Downloaded from central: https://repo.maven.apache.org/maven2/com/google/collections/google-collections/1.0/google-collections-1.0.jar (640 kB at 1.2 MB/s)
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to /project/user-app/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /project/user-app/src/main/java/dev/appsody/starter/StarterApplication.java:[8,5] <identifier> expected
[ERROR] /project/user-app/src/main/java/dev/appsody/starter/StarterApplication.java:[9,2] reached end of file while parsing
[INFO] 2 errors 
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  57.022 s
[INFO] Finished at: 2019-10-15T15:27:33Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project starter-app: Compilation failure: Compilation failure: 
[ERROR] /project/user-app/src/main/java/dev/appsody/starter/StarterApplication.java:[8,5] <identifier> expected
[ERROR] /project/user-app/src/main/java/dev/appsody/starter/StarterApplication.java:[9,2] reached end of file while parsing
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
subprocess exited with status 1
subprocess exited with status 1
error building at STEP "RUN mvn install -DskipTests": exit status 1
```

- When we checked the `build-push` container logs

`oc logs -f my-java-mp-webhook-1571152644-build-task-wndcp-pod-2f0e7d -c step-build-push`

Logs never returned anything which means the `build-push` step never ran since earlier step failed. :
```
[root@crofts1 common]# oc logs -f my-java-mp-webhook-1571152644-build-task-wndcp-pod-2f0e7d -c step-build-push
[root@crofts1 common]# 
```

--- a/packages/components/src/components/PipelineRun/PipelineRun.js
+++ b/packages/components/src/components/PipelineRun/PipelineRun.js
@@ -23,7 +23,8 @@ import {
   selectedTask,
   selectedTaskRun,
   stepsStatus,
-  taskRunStep
+  taskRunStep,
+  updateUnexecutedSteps
 } from '@tektoncd/dashboard-utils';
 
 import { Log, RunHeader, StepDetails, TaskTree } from '..';
@@ -245,7 +246,10 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
 
     const { definition, reason, status, stepName, stepStatus } = taskRunStep(
       selectedStepId,
-      taskRun
+      {
+        ...taskRun,
+        steps: updateUnexecutedSteps(taskRun.steps)
+      }
     );
 
     const logContainer = (

--- a/packages/components/src/components/Task/Task.js
+++ b/packages/components/src/components/Task/Task.js
@@ -17,6 +17,8 @@ import ChevronRight from '@carbon/icons-react/lib/chevron--right/16';
 import CloseFilled from '@carbon/icons-react/lib/close--filled/16';
 import { Spinner, Step } from '@tektoncd/dashboard-components';
 
+import { updateUnexecutedSteps } from '@tektoncd/dashboard-utils';
+
 import './Task.scss';
 
 class Task extends Component {
@@ -89,24 +91,26 @@ class Task extends Component {
         </a>
         {expanded && (
           <ol className="step-list">
-            {steps.map(({ id, reason: stepReason, status, stepName }) => {
-              const selected = selectedStepId === id;
-              const stepStatus =
-                reason === 'TaskRunCancelled' && status !== 'terminated'
-                  ? 'cancelled'
-                  : status;
-              return (
-                <Step
-                  id={id}
-                  key={id}
-                  onSelect={this.handleStepSelected}
-                  reason={stepReason}
-                  selected={selected}
-                  status={stepStatus}
-                  stepName={stepName}
-                />
-              );
-            })}
+            {updateUnexecutedSteps(steps).map(
+              ({ id, reason: stepReason, status, stepName }) => {
+                const selected = selectedStepId === id;
+                const stepStatus =
+                  reason === 'TaskRunCancelled' && status !== 'terminated'
+                    ? 'cancelled'
+                    : status;
+                return (
+                  <Step
+                    id={id}
+                    key={id}
+                    onSelect={this.handleStepSelected}
+                    reason={stepReason}
+                    selected={selected}
+                    status={stepStatus}
+                    stepName={stepName}
+                  />
+                );
+              }
+            )}
           </ol>
         )}
       </li>

--- a/packages/components/src/components/Task/Task.stories.js
+++ b/packages/components/src/components/Task/Task.stories.js
@@ -23,8 +23,8 @@ const props = {
 };
 
 const steps = [
-  { id: 'build', stepName: 'build' },
-  { id: 'test', stepName: 'test' }
+  { id: 'build', stepName: 'build', reason: 'Completed' },
+  { id: 'test', stepName: 'test', reason: 'Completed' }
 ];
 
 storiesOf('Task', module)

--- a/packages/components/src/components/Task/Task.test.js
+++ b/packages/components/src/components/Task/Task.test.js
@@ -40,6 +40,20 @@ it('Task renders steps in expanded state', () => {
   expect(queryByText(/a step/i)).toBeTruthy();
 });
 
+it('Task renders completed steps in expanded state', () => {
+  const steps = [
+    { id: 'step1', stepName: 'step 1', reason: 'Completed' },
+    { id: 'step2', stepName: 'step 2', reason: 'Error' },
+    { id: 'step3', stepName: 'step 3', reason: 'Completed' }
+  ];
+  const { queryByText } = renderWithIntl(
+    <Task {...props} expanded steps={steps} />
+  );
+  expect(queryByText(/step 1/i)).toBeTruthy();
+  expect(queryByText(/step 2/i)).toBeTruthy();
+  expect(queryByText(/step 3/i)).toBeTruthy();
+});
+
 it('Task renders success state', () => {
   renderWithIntl(<Task {...props} succeeded="True" />);
 });

--- a/packages/utils/src/utils/index.js
+++ b/packages/utils/src/utils/index.js
@@ -139,3 +139,33 @@ export function getStatusIcon({ reason, status }) {
 
   return Icon ? <Icon className="status-icon" /> : null;
 }
+
+// Update the status of steps that follow a step with an error
+export function updateUnexecutedSteps(steps) {
+  if (!steps) {
+    return steps;
+  }
+  let errorIndex = steps.length - 1;
+  return steps.map((step, index) => {
+    // Update errorIndex
+    if (step.reason !== 'Completed') {
+      errorIndex = Math.min(index, errorIndex);
+    }
+    // Update step
+    if (index > errorIndex) {
+      const s = {
+        ...step,
+        reason: '',
+        status: ''
+      };
+      if (step.stepStatus && step.stepStatus.terminated) {
+        s.stepStatus = {
+          ...step.stepStatus,
+          terminated: { ...step.stepStatus.terminated, reason: '' }
+        };
+      }
+      return s;
+    }
+    return step;
+  });
+}

--- a/packages/utils/src/utils/index.test.js
+++ b/packages/utils/src/utils/index.test.js
@@ -18,7 +18,8 @@ import {
   isRunning,
   selectedTask,
   stepsStatus,
-  taskRunStep
+  taskRunStep,
+  updateUnexecutedSteps
 } from '.';
 
 it('taskRunSteps with no taskRun', () => {
@@ -238,4 +239,72 @@ it('stepsStatus step is terminated with error', () => {
   expect(returnedStep.status).toEqual('terminated');
   expect(returnedStep.stepName).toEqual(stepName);
   expect(returnedStep.reason).toEqual(reason);
+});
+
+it('updateUnexecutedSteps no steps', () => {
+  const steps = [];
+  const wantUpdatedSteps = [];
+  const gotUpdatedSteps = updateUnexecutedSteps(steps);
+  expect(gotUpdatedSteps).toEqual(wantUpdatedSteps);
+});
+
+it('updateUnexecutedSteps undefined steps', () => {
+  let steps;
+  let wantUpdatedSteps;
+  const gotUpdatedSteps = updateUnexecutedSteps(steps);
+  expect(gotUpdatedSteps).toEqual(wantUpdatedSteps);
+});
+
+it('updateUnexecutedSteps no error steps', () => {
+  const steps = [
+    { reason: 'Completed', status: 'Terminated' },
+    {
+      reason: 'Running',
+      status: 'Unknown',
+      stepStatus: {}
+    }
+  ];
+  const wantUpdatedSteps = [...steps];
+  const gotUpdatedSteps = updateUnexecutedSteps(steps);
+  expect(gotUpdatedSteps).toEqual(wantUpdatedSteps);
+});
+
+it('updateUnexecutedSteps error step', () => {
+  const steps = [
+    {
+      reason: 'Completed',
+      status: 'Terminated',
+      stepStatus: { terminated: { reason: 'Completed' } }
+    },
+    {
+      reason: 'Error',
+      status: 'Error',
+      stepStatus: { terminated: { reason: 'Error' } }
+    },
+    {
+      reason: 'Completed',
+      status: 'Terminated',
+      stepStatus: { terminated: { reason: 'Completed' } }
+    }
+  ];
+  const wantUpdatedSteps = [
+    {
+      reason: 'Completed',
+      status: 'Terminated',
+      stepStatus: { terminated: { reason: 'Completed' } }
+    },
+    {
+      reason: 'Error',
+      status: 'Error',
+      stepStatus: { terminated: { reason: 'Error' } }
+    },
+    {
+      reason: '',
+      status: '',
+      stepStatus: { terminated: { reason: '' } }
+    }
+  ];
+
+  const gotUpdatedSteps = updateUnexecutedSteps(steps);
+  expect(gotUpdatedSteps).toEqual(wantUpdatedSteps);
 });

--- a/src/containers/TaskRun/TaskRun.js
+++ b/src/containers/TaskRun/TaskRun.js
@@ -24,7 +24,12 @@ import {
   StepDetails,
   TaskTree
 } from '@tektoncd/dashboard-components';
-import { getStatus, stepsStatus, taskRunStep } from '@tektoncd/dashboard-utils';
+import {
+  getStatus,
+  stepsStatus,
+  taskRunStep,
+  updateUnexecutedSteps
+} from '@tektoncd/dashboard-utils';
 
 import { fetchLogs } from '../../utils';
 
@@ -170,7 +175,10 @@ export /* istanbul ignore next */ class TaskRunContainer extends Component {
 
     const { definition, reason, status, stepName, stepStatus } = taskRunStep(
       selectedStepId,
-      taskRun
+      {
+        ...taskRun,
+        steps: updateUnexecutedSteps(taskRun.steps)
+      }
     );
 
     const logContainer = (

